gateway-server/src/main/java/org/apache/knox/gateway/websockets/JWTValidatorFactory.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.knox.gateway.websockets;
+
+import org.apache.knox.gateway.config.GatewayConfig;
+import org.apache.knox.gateway.i18n.messages.MessagesFactory;
+import org.apache.knox.gateway.provider.federation.jwt.JWTMessages;
+import org.apache.knox.gateway.provider.federation.jwt.filter.SignatureVerificationCache;
+import org.apache.knox.gateway.services.GatewayServices;
+import org.apache.knox.gateway.services.ServiceType;
+import org.apache.knox.gateway.services.security.token.TokenStateService;
+import org.apache.knox.gateway.services.security.token.impl.JWT;
+import org.apache.knox.gateway.services.security.token.impl.JWTToken;
+import org.apache.knox.gateway.services.topology.TopologyService;
+import org.apache.knox.gateway.topology.Service;
+import org.apache.knox.gateway.topology.Topology;
+import org.apache.knox.gateway.util.CertificateUtils;
+import org.eclipse.jetty.websocket.servlet.ServletUpgradeRequest;
+import javax.servlet.ServletException;
+import java.net.HttpCookie;
+import java.security.interfaces.RSAPublicKey;
+import java.text.ParseException;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class JWTValidatorFactory {
```

Comment:
Again, there is very little websocket specific in here and I think that the ServletUpgradeRequest can likely be made more generic depending on its hierarchy. It seems like this Factory should be more generic and common JWT code rather than websocket specific.

Reply:
if ServletUpgradeRequest can't be cast to or used to retrieve the HttpServletRequest to be more generic then maybe there should be an abstract based class that can override extractToken here.

Reply:
Unfortunately `ServletUpgradeRequest` is part of [Websocket implementation](https://www.eclipse.org/jetty/javadoc/jetty-9/org/eclipse/jetty/websocket/servlet/ServletUpgradeRequest.html) so difficult to make it generic. This idea is that factory is exposes a more generic `JWTValidator` class which is more generic and WS agnostic.

Reply:
`ServletUpgradeRequest `  is a special request, this is the point where HTTP to WS handoff takes place, hence a logical place to grab JWT. This makes it WS specific unfortunately.